File: src/main/java/rbq/wtf/lycoris/client/module/Modules/Render/HUD.java
package rbq.wtf.lycoris.client.module.modules.render;

import rbq.wtf.lycoris.client.Client;
import rbq.wtf.lycoris.client.event.EventTarget;
import rbq.wtf.lycoris.client.event.Render2DEvent;
import rbq.wtf.lycoris.client.gui.Font.FontLoaders;
import rbq.wtf.lycoris.client.module.Module;
import rbq.wtf.lycoris.client.module.ModuleCategory;
import rbq.wtf.lycoris.client.module.ModuleInfo;
import rbq.wtf.lycoris.client.value.BooleanValue;
import rbq.wtf.lycoris.client.value.ModeValue;
import rbq.wtf.lycoris.client.value.NumberValue;
import rbq.wtf.lycoris.client.wrapper.wrappers.wrapper.Minecraft;
import rbq.wtf.lycoris.client.wrapper.wrappers.wrapper.gui.ScaledResolution;

import java.awt.*;
import java.util.ArrayList;
import java.util.Comparator;

@ModuleInfo(name = "HUD", description = "Toggles visibility of the HUD.", category = ModuleCategory.Render, array = false)
public class HUD extends Module {
    public final BooleanValue waterMark = new BooleanValue("WaterMark", true);
    public final BooleanValue arrayList = new BooleanValue("ArrayList", true);
    public final ModeValue rainbowMode = new ModeValue("Rainbow Mode", new String[]{"Rainbow", "Astolfo", "Static", "StaticRainbow"}, 0);
    public final NumberValue rainbowSpeed = new NumberValue("RainbowSpeed", 100.0F, 0.0F, 2000.0F, 0.1F);

    @Override
    public void onEnable() {
    }

    @EventTarget
    public void onRender2D(Render2DEvent e) {
        ScaledResolution sc = new ScaledResolution(Minecraft.getMinecraft());
        if (waterMark.get()) {
            FontLoaders.default25.drawStringWithShadow("Dimples.love",
                    5,
                    5,
                    new Color(0, 200, 100).getRGB());
        }
        if (arrayList.get()) {
            ArrayList<Module> sorted = new ArrayList<>();
            for (Module m : Client.moduleManager.getModules()) {
                if (!m.getState()) continue;
                sorted.add(m);
            }
            sorted.sort(Comparator.comparingInt(mm -> -(FontLoaders.default16.getStringWidth(mm.getName()))));
            float posY = 0;
            for (Module m : sorted) {
                FontLoaders.default20.drawStringWithShadow(m.getName(),
                        sc.getScaledWidth() - FontLoaders.default20.getStringWidth(m.getName()) - 5,
                        posY + 5,
                        new Color(0, 200, 100).getRGB()
                );
                posY = posY + FontLoaders.default16.getStringHeight(m.getName()) + 4;
            }

        }
    }

}

File: src/main/java/rbq/wtf/lycoris/client/module/Modules/Render/HUD.kt
package rbq.wtf.lycoris.client.module.modules.render

import rbq.wtf.lycoris.client.Client
import rbq.wtf.lycoris.client.event.EventTarget
import rbq.wtf.lycoris.client.event.Render2DEvent
import rbq.wtf.lycoris.client.font.FontLoaders
import rbq.wtf.lycoris.client.module.Module
import rbq.wtf.lycoris.client.module.ModuleCategory
import rbq.wtf.lycoris.client.module.ModuleInfo
import rbq.wtf.lycoris.client.utils.ColorUtils
import rbq.wtf.lycoris.client.value.BooleanValue
import rbq.wtf.lycoris.client.value.ModeValue
import rbq.wtf.lycoris.client.value.NumberValue
import rbq.wtf.lycoris.client.wrapper.wrappers.Minecraft
import rbq.wtf.lycoris.client.wrapper.wrappers.gui.ScaledResolution
import java.awt.Color

@ModuleInfo(
    name = "HUD",
    description = "Toggles visibility of the HUD.",
    category = ModuleCategory.Render
)
class HUD : Module() {
    private val arrayList = BooleanValue("ArrayList", true)
    private val waterMark = BooleanValue("WaterMark", true)
    private val rainbow = BooleanValue("Rainbow", true)
    private val waterMarkMode = ModeValue("WaterMark Mode", arrayOf("LOGO", "MAGIC"), 0)
    private val rainbowMode = ModeValue("Rainbow Mode", arrayOf("Rainbow", "Astolfo", "Static", "StaticRainbow"), 0)
    private val rainbowSpeed = NumberValue("RainbowSpeed", 10.0f, 0.0f, 200.0f, 5f)
    override fun onEnable() {}

    @EventTarget
    fun onRender2D(e: Render2DEvent?) {
        val sc = ScaledResolution(Minecraft.minecraft)
        if (waterMark.get()) {
            val text = if (waterMarkMode.get() == 0) "Dimples.love" else ColorUtils.randomMagicText("Dimples.love")
            FontLoaders.default30.drawStringWithShadow(
                text,
                5.0,
                5.0,
                Color(0, 200, 100).rgb
            )
        }
        if (arrayList.get()) {
            val sorted = Client.moduleManager.modules.filter { m -> (m.state && m.array) }
                .sortedBy { -FontLoaders.default16.getStringWidth(it.name) }
            var posY = 0f
            var rainbowOffset = 0L
            for (m in sorted) {
                val color = if (rainbow.get()) ColorUtils.rainbow(rainbowOffset) else Color(0, 200, 100)
                FontLoaders.default20.drawStringWithShadow(
                    m.name,
                    (sc.scaledWidth - FontLoaders.default20.getStringWidth(m.name) - 5).toDouble(),
                    (posY + 5).toDouble(),
                    color.rgb
                )
                posY += FontLoaders.default16.getStringHeight(m.name) + 4
                rainbowOffset += rainbowSpeed.get().toLong() * 100
            }
        }
    }
}